## ДЗ2. Тема “Создание признакового пространства” - пока не доделано

In [21]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
import seaborn as sns
import nltk
import warnings 
warnings.filterwarnings("ignore", category=DeprecationWarning)
import os
import pickle

In [2]:

df2 = pd.read_pickle("./words2.pkl")
df2.head()

id  label                                              tweet  \
0   1    0.0   @user when a father is dysfunctional and is s...   
1   2    0.0  @user @user thanks for #lyft credit i can't us...   
2   3    0.0                                bihday your majesty   
3   4    0.0  #model   i love u take with u all the time in ...   
4   5    0.0             factsguide: society now    #motivation   

                                         clean_tweet  \
0  when father is dysfunctional and is so selfish...   
1  thanks for lyft credit cannot use cause they d...   
2                                bihday your majesty   
3    model love you take with you all the time in ur   
4                  factsguide society now motivation   

                                         tweet_token  \
0  [when, father, is, dysfunctional, and, is, so,...   
1  [thanks, for, lyft, credit, can, not, use, cau...   
2                            [bihday, your, majesty]   
3  [model, love, you, take, with, you, all, the, ...   
4             [factsguide, society, now, motivation]   

                                tweet_token_filtered  \
0  [father, dysfunctional, selfish, drags, kids, ...   
1  [thanks, lyft, credit, use, cause, offer, whee...   
2                                  [bihday, majesty]   
3                      [model, love, take, time, ur]   
4                  [factsguide, society, motivation]   

                                       tweet_stemmed  \
0  [father, dysfunct, selfish, drag, kid, dysfunc...   
1  [thank, lyft, credit, use, caus, offer, wheelc...   
2                                  [bihday, majesti]   
3                      [model, love, take, time, ur]   
4                        [factsguid, societi, motiv]   

                                    tweet_lemmatized  
0  [father, dysfunctional, selfish, drag, kid, dy...  
1  [thanks, lyft, credit, use, cause, offer, whee...  
2                                  [bihday, majesty]  
3                      [model, love, take, time, ur]  
4                  [factsguide, society, motivation]

1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
●	Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
●	Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
●	Исключим стоп-слова с помощью stop_words='english'.
●	Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().


In [15]:
#df2 = combine_df.copy()
#df2.to_pickle("./words.pkl")

In [16]:
df2.head(2)

id  label                                              tweet  \
0   1    0.0   @user when a father is dysfunctional and is s...   
1   2    0.0  @user @user thanks for #lyft credit i can't us...   

                                         clean_tweet  \
0  when father is dysfunctional and is so selfish...   
1  thanks for lyft credit cannot use cause they d...   

                                         tweet_token  \
0  [when, father, is, dysfunctional, and, is, so,...   
1  [thanks, for, lyft, credit, can, not, use, cau...   

                                tweet_token_filtered  \
0  [father, dysfunctional, selfish, drags, kids, ...   
1  [thanks, lyft, credit, use, cause, offer, whee...   

                                       tweet_stemmed  \
0  [father, dysfunct, selfish, drag, kid, dysfunc...   
1  [thank, lyft, credit, use, caus, offer, wheelc...   

                                    tweet_lemmatized  
0  [father, dysfunctional, selfish, drag, kid, dy...  
1  [thanks, lyft, credit, use, cause, offer, whee...

In [17]:
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd

count_vectorizer = CountVectorizer(max_df = 0.9, stop_words='english', max_features = 1000)

In [18]:
tweet_stemmed = df2['tweet_stemmed'].apply(lambda x: ' '.join(x))
bow_stemmed_tweet = count_vectorizer.fit_transform(tweet_stemmed)
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(bow_stemmed_tweet.toarray(), columns = feature_names).head(2)

MemoryError: Unable to allocate 375. MiB for an array with shape (49159, 1000) and data type int64

In [7]:
s1 = df2['tweet_stemmed'][0]
s2 = df2['tweet_lemmatized'][0]
for i in range(1, df2.shape[0]):
    s1.extend(df2['tweet_stemmed'][i])
    s2.extend(df2['tweet_lemmatized'][i])
    #s1 += df2['tweet_stemmed'][i]
    #s2 += df2['tweet_lemmatized'][i]
s1 = list(set(s1))
s2 = list(set(s2))
print(len(s1), len(s2))

41057 45922


In [19]:
# Пожоже не получилось, не 1000 фичей, а все показывает.

# Создаем the Bag-of-Words модель
bag_tweet_stemmed = count_vectorizer.fit_transform(s1)
#bag_tweet_lemmatized = count_vectorizer.fit_transform(s2)

# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(bag_tweet_stemmed.toarray(), columns = feature_names)

MemoryError: Unable to allocate 313. MiB for an array with shape (41057, 1000) and data type int64

In [9]:
# Создаем the Bag-of-Words модель
bag_tweet_lemmatized = count_vectorizer.fit_transform(s2)
# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(bag_tweet_lemmatized.toarray(), columns = feature_names)

abbotabad  aching  acousticguitar  acre  acted  adamlambe  adc  \
0              0       0               0     0      0          0    0   
1              0       0               0     0      0          0    0   
2              0       0               0     0      0          0    0   
3              0       0               0     0      0          0    0   
4              0       0               0     0      0          0    0   
...          ...     ...             ...   ...    ...        ...  ...   
45917          0       0               0     0      0          0    0   
45918          0       0               0     0      0          0    0   
45919          0       0               0     0      0          0    0   
45920          0       0               0     0      0          0    0   
45921          0       0               0     0      0          0    0   

       aejazsaiyed  affraid  aftar  ...  yehvaadaraha  yippe  yojanas  \
0                0        0      0  ...             0      0        0   
1                0        0      0  ...             0      0        0   
2                0        0      0  ...             0      0        0   
3                0        0      0  ...             0      0        0   
4                0        0      0  ...             0      0        0   
...            ...      ...    ...  ...           ...    ...      ...   
45917            0        0      0  ...             0      0        0   
45918            0        0      0  ...             0      0        0   
45919            0        0      0  ...             0      0        0   
45920            0        0      0  ...             0      0        0   
45921            0        0      0  ...             0      0        0   

       yorkshire  youngecoengineers  youuu  yql  yummieee  yxeshame  zadiq  
0              0                  0      0    0         0         0      0  
1              0                  0      0    0         0         0      0  
2              0                  0      0    0         0         0      0  
3              0                  0      0    0         0         0      0  
4              0                  0      0    0         0         0      0  
...          ...                ...    ...  ...       ...       ...    ...  
45917          0                  0      0    0         0         0      0  
45918          0                  0      0    0         0         0      0  
45919          0                  0      0    0         0         0      0  
45920          0                  0      0    0         0         0      0  
45921          0                  0      0    0         0         0      0  

[45922 rows x 1000 columns]

2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
●	Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
●	Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
●	Исключим стоп-слова с помощью stop_words='english'.
●	Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().


In [10]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [11]:
tfidf_vectorizer = TfidfVectorizer(max_df = 0.9, stop_words='english', max_features = 1000)

In [12]:
values_stemmed = tfidf_vectorizer.fit_transform(s1)

# Show the Model as a pandas DataFrame
feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(values_stemmed.toarray(), columns = feature_names)

MemoryError: Unable to allocate 313. MiB for an array with shape (41057, 1000) and data type float64

In [ ]:
values_lemmatized = tfidf_vectorizer.fit_transform(s2)

# Show the Model as a pandas DataFrame
feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(values_lemmatized.toarray(), columns = feature_names)

3. Натренируем gensim.models.Word2Vec модель на наших данных.
●	Тренировать будем на токенизированных твитах combine_df['tweet_token']
●	Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
●	Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.


In [3]:
from gensim.models import Word2Vec

In [4]:
df_tweet_token = df2['tweet_token'].tolist()



In [5]:
w2v_model = Word2Vec( df_tweet_token, size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 3, seed = 34)

In [7]:
w2v_model.train(df_tweet_token, total_examples=len(df_tweet_token), epochs=20)

(9141720, 11726520)

4. Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

In [8]:
# Как то совсем не в тему вышло
w2v_model.most_similar(positive=['dinner'])

[('bolognese', 0.5677131414413452),
 ('bihdaydinner', 0.5553514957427979),
 ('spaghetti', 0.5419812202453613),
 ('cookout', 0.5407297611236572),
 ('shawarma', 0.5363484621047974),
 ('burritos', 0.5319509506225586),
 ('saturdate', 0.5311754941940308),
 ('enroute', 0.528834879398346),
 ('tacotuesday', 0.5284525752067566),
 ('whoopppp', 0.5270119309425354)]

In [9]:
w2v_model.most_similar(positive=['trump'])

[('donald', 0.55143141746521),
 ('suppoer', 0.526889443397522),
 ('conman', 0.5234178304672241),
 ('donaldtrump', 0.5186693072319031),
 ('dumptrump', 0.5072817802429199),
 ('impeachment', 0.5014523267745972),
 ('unfavorability', 0.49996960163116455),
 ('melo', 0.49718108773231506),
 ('landslide', 0.4966769218444824),
 ('fuhered', 0.49508601427078247)]

5. Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
Давайте проверим векторное представление любого слова из нашего корпуса, например "food".


In [10]:
w2v_model['food']

array([ 3.43487054e-01, -6.04504347e-01, -2.13789538e-01, -3.90901119e-01,
       -2.03109324e-01,  3.21404517e-01, -9.93711501e-02,  9.52136219e-02,
       -6.53275177e-02, -5.14910936e-01,  5.35589039e-01,  5.16326606e-01,
        1.51602983e-01,  5.87235987e-01,  2.04601251e-02,  2.75645703e-01,
        2.89819688e-01, -3.13151687e-01, -2.58644044e-01, -3.73612009e-02,
        1.82014719e-01,  5.83277978e-02, -5.05585432e-01,  8.10861349e-01,
        7.70480454e-01,  4.95931387e-01, -2.27258224e-02,  4.79464471e-01,
       -1.30772427e-01, -2.94187397e-01,  4.27245349e-01,  8.24302360e-02,
        3.16935964e-02,  6.07073426e-01,  7.12193176e-02, -1.36056229e-01,
        6.30167052e-02,  1.36019126e-01, -7.03791797e-01, -7.41061032e-01,
        2.27099210e-02,  6.35671735e-01,  1.39485493e-01, -7.17239827e-02,
        1.68617144e-01,  2.93544948e-01, -7.67298877e-01, -2.46592969e-01,
        6.61781192e-01,  2.69126207e-01,  5.44183135e-01, -3.60186696e-01,
        1.03131562e-01, -

6. Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.
Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size))
и поделить финальный вектор на количество слов в твите.
На выходе должен получиться wordvec_df.shape = (49159, 200).


In [11]:
def tws2vec(tweets_list, size=200):
    num = len(tweets_list)
    vec = np.zeros((1, size))
    if num == 0:
        return vec
    for word in tweets_list:
        try:
            vec += w2v_model[word].reshape((1, size))
        except KeyError:
            continue
    return vec / num



In [12]:
twsvec_df = df2['tweet_token'].apply(tws2vec)
twsvec_df = np.concatenate(twsvec_df)

In [13]:
twsvec_df.shape

(49159, 200)

In [24]:
with open('twsvec.pkl', 'wb') as f:
    pickle.dump(twsvec_df, f)
